Move std::span-related utility functions to <wtf/StdLibExtras.h>

cdumez opened the pull request:
#### 8c35e078b772cb94d9436eef8a90a33a2c718088
<pre>
Move std::span-related utility functions to &lt;wtf/StdLibExtras.h&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=272653">https://bugs.webkit.org/show_bug.cgi?id=272653</a>

Reviewed by Darin Adler and Sam Weinig.

* Source/WTF/wtf/Algorithms.h:
(WTF::spanReinterpretCast): Deleted.
(WTF::sizeof): Deleted.
(WTF::equalSpans): Deleted.
(WTF::memcpySpan): Deleted.
(WTF::memsetSpan): Deleted.
* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanReinterpretCast):
(WTF::sizeof):
(WTF::equalSpans):
(WTF::memcpySpan):
(WTF::memsetSpan):
* Source/WTF/wtf/persistence/PersistentCoders.cpp:
* Source/WTF/wtf/persistence/PersistentCoders.h:
* Source/WTF/wtf/text/Base64.cpp:
* Source/WTF/wtf/text/WTFString.cpp:
* Source/WebCore/PAL/pal/text/win/TextCodecWin.cpp:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
* Source/WebCore/dom/TextEncoder.cpp:
* Source/WebCore/platform/generic/KeyedEncoderGeneric.cpp:
* Source/WebCore/platform/sql/SQLiteStatement.cpp:
* Source/WebCore/workers/ScriptBuffer.cpp:
* Source/WebKit/Platform/IPC/DaemonCoders.h:
* Source/WebKit/Platform/IPC/Decoder.h:
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
* Source/WebKit/UIProcess/DeviceIdHashSaltStorage.cpp:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:

* Tools/TestWebKitAPI/glib/CMakeLists.txt:
Stop forcing build with C++17 since std::span requires C++20.

Canonical link: <a href="https://commits.webkit.org/277480@main">https://commits.webkit.org/277480@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b1e7be3f3ff787a1383793b2a1b20454d3c388d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47746 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26935 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50511 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50429 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43801 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50053 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32792 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24416 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48328 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24589 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41225 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20156 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22070 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42416 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5795 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41028 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44093 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42839 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52323 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47237 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22782 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19112 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46165 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24054 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45200 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24842 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54737 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6751 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23774 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11237 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -671,6 +671,51 @@ template<typename OptionalType> auto valueOrDefault(OptionalType&& optionalValue
+template<typename T, std::size_t Extent1, std::size_t Extent2>
+bool equalSpans(std::span<T, Extent1> a, std::span<T, Extent2> b)
+{
+    if (a.size() != b.size())
```
You could make an immediate path (or potentially just disallow with a static_assert) for when Extent1 != Extent2.

```c++
if constexpr (Extent1 != Extent2) {
    return false
} else {
    if (a.size() != b.size())
        return false;
    return !memcmp(a.data(), b.data(), a.size() * sizeof(T));
}
```

cdumez replied:
This doesn't seem right? What if `Extent1` is `std::dynamic_extent` and `Extent2` is 4, the spans could still be equal.

cdumez replied:
Also, I think is currently very efficient for spans with non-dynamic extent given the implementation of `size()` for such types:
```
constexpr size_type size() const noexcept { return _Extent; }
```

weinig replied:
> This doesn't seem right? What if `Extent1` is `std::dynamic_extent` and `Extent2` is 4, the spans could still be equal.

Yeah, I guess you would have to check that they were not dynamic extent, so:

You could make an immediate path (or potentially just disallow with a static_assert) for when Extent1 != Extent2.

```c++
if constexpr (Extent1 != Extent2 && Extent1 != std::dynamic_extent && Extent2 != std::dynamic_extent) {
    return false
} else {
    if (a.size() != b.size())
        return false;
    return !memcmp(a.data(), b.data(), a.size() * sizeof(T));
}
```

But yeah, it would be unlikely to be hit or worth it.

weinig commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -671,6 +671,51 @@ template<typename OptionalType> auto valueOrDefault(OptionalType&& optionalValue
+{
+    if (a.size() != b.size())
+        return false;
+    return !memcmp(a.data(), b.data(), a.size() * sizeof(T));
```
std::span, for whatever reason, has a function for the size of the byte sequence, `std::span:: size_bytes()` and you seem to use it in most other functions here.

weinig commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -671,6 +671,51 @@ template<typename OptionalType> auto valueOrDefault(OptionalType&& optionalValue
+template<typename T, std::size_t Extent>
+void memsetSpan(std::span<T, Extent> destination, uint8_t byte)
+{
+    static_assert(std::is_trivially_copyable_v<T>);
```
Is `std::is_trivially_copyable_v` enough to cover a types with non-trivial destructors or constructors?

cdumez replied:
Per https://en.cppreference.com/w/cpp/language/classes#Trivially_copyable_class:
- each eligible copy constructor is trivial
- has a non-deleted trivial destructor

https://en.cppreference.com/w/cpp/string/byte/memset also mentions that behavior for non-trivially copyable types is undefined.

weinig commented on Tools/TestWebKitAPI/glib/CMakeLists.txt:
```diff
@@ -176,7 +176,6 @@ macro(ADD_WPE_QT_TEST test_name)
     set_target_properties(${test_name} PROPERTIES
         RUNTIME_OUTPUT_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/TestWebKitAPI/WPEQt
         AUTOMOC ON
-        CXX_STANDARD 17
```
This is perhaps worth calling out in the change log if intentional.

cdumez replied:
Yes, it's intentional. std::span is C++20 only and some of the files in this test include StdLibExtras.h. I'll mention it in the commit log.

cdumez committed:
Move std::span-related utility functions to <wtf/StdLibExtras.h>
https://bugs.webkit.org/show_bug.cgi?id=272653

Reviewed by Darin Adler and Sam Weinig.

* Source/WTF/wtf/Algorithms.h:
(WTF::spanReinterpretCast): Deleted.
(WTF::sizeof): Deleted.
(WTF::equalSpans): Deleted.
(WTF::memcpySpan): Deleted.
(WTF::memsetSpan): Deleted.
* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanReinterpretCast):
(WTF::sizeof):
(WTF::equalSpans):
(WTF::memcpySpan):
(WTF::memsetSpan):
* Source/WTF/wtf/persistence/PersistentCoders.cpp:
* Source/WTF/wtf/persistence/PersistentCoders.h:
* Source/WTF/wtf/text/Base64.cpp:
* Source/WTF/wtf/text/WTFString.cpp:
* Source/WebCore/PAL/pal/text/win/TextCodecWin.cpp:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
* Source/WebCore/dom/TextEncoder.cpp:
* Source/WebCore/platform/generic/KeyedEncoderGeneric.cpp:
* Source/WebCore/platform/sql/SQLiteStatement.cpp:
* Source/WebCore/workers/ScriptBuffer.cpp:
* Source/WebKit/Platform/IPC/DaemonCoders.h:
* Source/WebKit/Platform/IPC/Decoder.h:
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
* Source/WebKit/UIProcess/DeviceIdHashSaltStorage.cpp:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:

* Tools/TestWebKitAPI/glib/CMakeLists.txt:
Stop forcing build with C++17 since std::span requires C++20.

Canonical link: https://commits.webkit.org/277480@main
